pkg/querier/frontend/frontend_test.go

```diff
@@ -251,8 +255,75 @@ func TestFrontendCheckReady(t *testing.T) {
 	}
 }
 
-func testFrontend(t *testing.T, config Config, handler http.Handler, test func(addr string), matchMaxConcurrency bool) {
+func TestFrontend_LogsSlowQueriesFormValues_Issue3111(t *testing.T) {
+	// Create an HTTP server listening locally. This server mocks the downstream
+	// Prometheus API-compatible server.
+	downstreamListen, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	downstreamServer := http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			_, err := w.Write([]byte(responseBody + responseBody + responseBody + responseBody + responseBody))
+			require.NoError(t, err)
+		}),
+	}
+
+	defer downstreamServer.Shutdown(context.Background()) //nolint:errcheck
+	go downstreamServer.Serve(downstreamListen)           //nolint:errcheck
+
+	// Configure the query-frontend with the mocked downstream server.
+	config := defaultFrontendConfig()
+	config.LogQueriesLongerThan = 1 * time.Microsecond
+	config.DownstreamURL = fmt.Sprintf("http://%s", downstreamListen.Addr())
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Do we need this? Given we're testing the case of the downstream URL, is this required at all? I found it confusing having both `w.Write([]byte(responseBody + responseBody + responseBody + responseBody + responseBody))` here and in the `downstreamServer`.

Reply:
Removed handler as it is no longer needed